### PR TITLE
Test random usage of peerset api

### LINKED ISSUE #2200: Fuzz the peerset manager to test coherency
Write a test for the `peerset` crate that uses the API in a random way, and verifies if the output is coherent (eg. if the peerset doesn't send "dropped" twice in a row for the same node).

cc @debris

### PR DESCRIPTION
**Pull request by gterzian**
FIX #2200

### CODE REVIEW

**Comment by gterzian** [comment]
@tomaka One issue I ran into is that one can't keep exactly track of the "last message" since some messages come with a `PeerId`, and others with a `IncomingIndex`. 

Would there be a way to figure out one from the other?

**Comment by tomaka** [comment]
Will properly review later, but I don't think we should change the API because of tests.
Why not just have a hashmap that associates incoming indices with peerids?

Also, since it's pretty large and only deals with the public API, I suggest we move the code that this PR adds into the `tests/` directory instead.

**Comment by gterzian** [comment]
> Why not keep a hashmap that associates incoming indices with peerids?

When a message comes in with an indice, how to associate it with a peer id?

> I suggest we move the code that this PR adds into the tests/ directory instead

Ok

**Comment by tomaka** [comment]
Oh I didn't realize you were only dumping a lot of messages into the peerset then reading the results.
That's not a good enough coverage in my opinion. We should simulate an actual network and react to what the peerset sends to us and in a random way. It really isn't trivial to do unfortunately.

**Comment by gterzian** [comment]
Ok so re the index vs id, they can be associated them when setting things up. 

> We should simulate an actual network and react to what the peerset sends to us and in a random way.

Why do we need a network setup? It seems easier to test the peerset as a blackbox by randomly using the API, even in ways that currently isn't used as part of an actually running network. 

For example, from the current test failure it appears to be quite easy to have an `Accept` message following a `Connect` message for a peer by doing a `dropped` followed by an `incoming`. So perhaps that is not the intended usage of the API, and "it will never happen". Yet now we have the information that it can happen as a result of random usage of the API. 

We can then either change our expectation, and say that it's OK, or change the peerset so that the expectation holds even in that specific case.

**Comment by tomaka** [comment]
> Why do we need a network setup?

By "simulate an actual network", I mean "use the API to make the peerset believe that we have an actual network".

> it appears to be quite easy to have an Accept message following a Connect message for a peer by doing a dropped followed by an incoming.

"Connect -> dropped -> incoming -> Accept" is indeed a valid API usage.
Unfortunately figuring out what is correct and incorrect is quite hard, both in terms of code and in terms of intuition.

**Comment by gterzian** [comment]
Ok I've added some checks that can actually express an expectations of a sequence of actions like "Connect -> dropped -> incoming -> Accept". 

It seems that " dropped -> incoming -> Accept" holds so far, but that serie comes out often without first having received a `Connect` message. 

>  "use the API to make the peerset believe that we have an actual network"

It's already pretty encouraging that using the API randomly seems to result in mostly expect output in terms of messages. If we focus more on using the API in a way that an actual network would, are we not going to exclude weird unexpected cases? The "random" input could be adapted to not be completely random and match what would otherwise be expected...

**Comment by gavofyork** [comment]
@tomaka @gterzian how is this looking? will it be in a mergeable state any time soon?

**Comment by gterzian** [comment]
Currently tests are failing on the assertions of "Connect -> dropped -> incoming -> Accept" always being received in that order. If we reduce this to only "dropped -> incoming -> Accept", they would pass. 

Other then that technicality, we need to decide on an overall approach. 

I am in favor of the approach taken here, which is to send completely random messages to the peer set as input, and then make assertions on the output we get from the peerset, ensuring that no matter the sequence of input, certain invariant hold on the output. 

@tomaka I think you wanted to take a different approach? Do you have suggestions on steps that could be taken in that direction?

**Comment by tomaka** [comment]
> I think you wanted to take a different approach?

I'm fine with the approach in this PR, but I think it's insufficient. We should adjust the messages we send to the PSM based on what it is sending back to us.

I opened #2470 with that other approach, but we could merge both.

**Comment by gterzian** [comment]
Ok, I'd taken a look and I think it wouldn't hurt to have both tests run side by side. 

I've also fixed a logical error in the assertions on this one, it now passes. 

The folder introduced here is named `test`, while the other PR introduces a `tests`, but I'm hopeful that's a difference we can overcome...

File: core/peerset/src/test/mod.rs [hunk 1]
```diff
@@ -225,8 +225,6 @@ fn test_random_api_use() {
                         match (drop_position, incoming_position) {
                             (Some(drop), Some(incoming)) => {
                                 assert!(drop < incoming);
-                                println!("Last messages: {:?}", last_messages);
-                                assert!(last_messages.contains(&Message::Connect(peer_id.clone())));
```
Comment (gterzian) [inline core/peerset/src/test/mod.rs]: Since the `Connect` message has already been removed from the `VecDeque` above, this was a faulty assertion...

**Comment by tomaka** [comment]
> The folder introduced here is named test, while the other PR introduces a tests, but I'm hopeful that's a difference we can overcome...

I went for an external test since we test the API surface of the crate, as opposed to an internal API.

**Comment by gterzian** [comment]
Ok, I'd be happy to rebase after the other PR merges, and just add this test to the same file. 

Fixed a couple of bugs in this test. Perhaps you could also take a look that the main assertions, about the order of operations, make sense? 

I'll look more into the other PR, and see if I can dig up the reason for the current failure...

**Comment by tomaka** [comment]
> I'll look more into the other PR, and see if I can dig up the reason for the current failure...

It's entirely possible that there's an issue in the PSM. It wouldn't be a bad idea to test this PR against #2440.

**Comment by gavofyork** [comment]
@gabreal this looks stale - should it be closed?

**Comment by gavofyork** [comment]
@gterzian this looks stale - should it be closed?

**Comment by gterzian** [comment]
Ok, closing because I think https://github.com/paritytech/substrate/pull/2470 gives us the coverage we want...

**Pull request closed without merging**